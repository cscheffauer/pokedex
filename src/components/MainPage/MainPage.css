@font-face {
    font-display: swap;
    font-family: 'STAR JEDI';
    font-style: normal;
    font-weight: normal;
    src: local('STAR JEDI'), url('../../fonts/Starjedi.ttf') format('woff');
}

h1 {
    font-family: 'STAR JEDI';
    color: white;
}