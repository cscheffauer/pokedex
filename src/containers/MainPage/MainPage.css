html {
    scroll-behavior: smooth;
}
body {
    background-color: #181B1D;
}

@font-face {
    font-display: swap;
    font-family: 'PTSans';
    font-style: normal;
    font-weight: normal;
    src: local('PTSans'), url('../../fonts/PTSans-Regular.ttf') format('woff');
}


.wrapper {
    height: 100%;
    width: 100%;
    display: flex;
    flex-direction: row;
    justify-content: space-around;
}

.main {
    background-color: #BFBFBF;
    width: 865px;
    margin-top: 50px;
    height: calc(100vh - 50px);
    border-top-right-radius: 50px;
    border-top-left-radius: 50px;
    position: absolute;
}

.pokeball {
    width: 175px;
    height: 175px;
    position: absolute;
    top: -30px;
    right: -25px;
}

.pokeballHover:hover {
    cursor: pointer;
}





/* Small devices (portrait tablets and large phones, 600px and up) 
@media only screen and (min-width: 600px) {...}

 Medium devices (landscape tablets, 768px and up) 
@media only screen and (min-width: 768px) {...}

 Large devices (laptops/desktops, 992px and up) 
@media only screen and (min-width: 992px) {...}

 Extra large devices (large laptops and desktops, 1200px and up) 
@media only screen and (min-width: 1200px) {...} */