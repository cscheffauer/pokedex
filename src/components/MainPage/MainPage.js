import React, { Component } from 'react';   //destructuring (instead of React.Component)

import CardList from '../CardList';
import SearchBox from '../SearchBox';
import Scroll from '../Scroll';
import ErrorBoundry from '../ErrorBoundry';
import Header from '../Header';
import './MainPage.css'; //to import the appropriate css file for this component


class MainPage extends Component {   //class App will use the component lib / Component expects that there is a render function in this class

    componentDidMount() {
        this.props.onRequestPeople();
    }
    filterPeople = () => {
        return this.props.people.filter(person => {      //filter the people and return them to filteredpeople
            return person.name.toLowerCase().includes(this.props.searchField.toLowerCase());
        });
    }


    render() {
        const { onSearchChange, firstPeoplePending } = this.props;

        return firstPeoplePending ?  //return html back depending on the isPending flag
            <h1 className='tc'>Loading</h1>
            :
            <div className='tc' >
                <Header />
                <SearchBox searchChange={onSearchChange} />
                <Scroll>
                    <ErrorBoundry>
                        <CardList people={this.filterPeople()} />
                    </ErrorBoundry>
                </Scroll>
            </div>
    }
}
export default MainPage;

//connect is a higher order function - will return another function -> that's why it is written that way above
//
//connect() is subscribed to the redux store now and it accepts 2 params: 
//1st is mapStateToProps: what state the component is interested in
//2nd is mapDispatchToProps: 