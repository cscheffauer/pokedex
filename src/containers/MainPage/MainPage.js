import React, { Component, Suspense, lazy, Fragment } from 'react';   //destructuring (instead of React.Component)

import {
    INITIAL_ROUTE,
    IMPRINT_ROUTE,
    TECHNOLOGYINFO_ROUTE
} from '../../constants'       //get constants form constants file

import CardList from '../../components/CardList';
import SearchBox from '../../components/SearchBox';
import Scroll from '../../components/Scroll';
import ErrorBoundry from '../../components/ErrorBoundry';
import LoadingSpinner from '../../components/Layout/LoadingSpinner/LoadingSpinner';
import TechnologyInfoButton from '../../components/Layout/TechnologyInfoButton/TechnologyInfoButton';


import './MainPage.css'; //to import the appropriate css file for this component
import ImprintButton from '../../components/Layout/ImprintButton/ImprintButton';


const AsyncPageTechnologyInfo = lazy(() => import('../../components/Pages/TechnologyInfo/TechnologyInfo'));
const AsyncPageImprint = lazy(() => import('../../components/Pages/Imprint/Imprint'));

class MainPage extends Component {   //class App will use the component lib / Component expects that there is a render function in this class

    constructor() {
        super();
        this.state = {
            searchBoxInput: "",
            scrollContainer: React.createRef(),
            searchBoxRef: React.createRef(),
        }

    }

    setSearchChange = (searchBoxInput) => {
        this.setState({ searchBoxInput: searchBoxInput });
    }

    getRouteComponent = (route) => {

        if (route === INITIAL_ROUTE) {
            return (
                <Fragment>
                    <SearchBox searchBoxRef={this.state.searchBoxRef} searchBoxInput={this.state.searchBoxInput} setSearchChange={this.setSearchChange} />
                    <Scroll scrollContainer={this.state.scrollContainer}>
                        <ErrorBoundry>
                            {
                                <CardList searchBoxRef={this.state.searchBoxRef} searchBoxInput={this.state.searchBoxInput} setSearchChange={this.setSearchChange} />
                            }
                        </ErrorBoundry>
                    </Scroll>
                </Fragment>
            )
        } else if (route === IMPRINT_ROUTE) {
            return (
                <Suspense fallback={<LoadingSpinner swapiLabel={false} />}>
                    <AsyncPageImprint />
                </Suspense>
            )
        } else if (route === TECHNOLOGYINFO_ROUTE) {
            return (
                <Suspense fallback={<LoadingSpinner swapiLabel={false} />}>
                    <AsyncPageTechnologyInfo />
                </Suspense>
            )
        }
    }

    render() {
        const { route, onRouteChange } = this.props;
        let page = this.getRouteComponent(route);

        const setInitialRoute = () => {
            (route !== INITIAL_ROUTE) && onRouteChange(INITIAL_ROUTE);
        }

        return (
            <>
                <div className='wrapper'>
                    <div className='main'>
                        {page}
                        <img onClick={() => setInitialRoute()} alt='pokeball' className={`pokeball ` + ((route !== INITIAL_ROUTE) ? `pokeballHover` : ``)} src="images/pokeball_grey.png" />
                    </div>
                </div>
                <TechnologyInfoButton onRouteChange={onRouteChange} />
                <ImprintButton onRouteChange={onRouteChange} />
            </>
        )
    }
}
export default MainPage;
