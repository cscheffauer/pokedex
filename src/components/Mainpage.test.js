import { shallow, mount, render } from 'enzyme';
import React from 'react';          //need to import react to use JSX in the shallow
import MainPage from './MainPage/MainPage';

let wrapper, wrapper2, wrapper3, wrapper4;

beforeEach(() => {
    const mockProps = {
        onRequestRobots: jest.fn(),
        robots: [],
        searchField: '',
        isPending: false
    }
    wrapper = shallow(<MainPage {...mockProps} />);
})

it('renders the mainpage without crashing', () => {
    expect(wrapper).toMatchSnapshot();
});

it('filters robots correctly', () => {
    expect(wrapper.instance().filterRobots()).toEqual([]);          //check if filtering of a empty robots array with a empty searchfield will return an empty array

    const mockProps2 = {
        onRequestRobots: jest.fn(),
        robots: [{
            id: 3,
            name: 'John',
            email: 'john@gmail.com'
        }],
        searchField: 'John',
        isPending: false
    }
    wrapper2 = shallow(<MainPage {...mockProps2} />);
    expect(wrapper2.instance().filterRobots()).toEqual([{                   //check if filtering the robots array will return the array we searched for
        id: 3,
        name: 'John',
        email: 'john@gmail.com'
    }]);
})

it('filters robots correctly 2', () => {
    const mockProps3 = {
        onRequestRobots: jest.fn(),
        robots: [{
            id: 3,
            name: 'John',
            email: 'john@gmail.com'
        }],
        searchField: 'a',
        isPending: false
    }
    const filteredRobots = [];
    wrapper3 = shallow(<MainPage {...mockProps3} />);
    expect(wrapper3.instance().filterRobots()).toEqual(filteredRobots);
})

it('returns isPending correctly', () => {
    const mockProps4 = {
        onRequestRobots: jest.fn(),
        robots: [],
        searchField: '',
        isPending: true
    }
    wrapper4 = render(<MainPage {...mockProps4} />);
    expect(wrapper4).toMatchSnapshot();
})