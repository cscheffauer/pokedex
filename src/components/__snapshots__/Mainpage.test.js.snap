// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the mainpage without crashing 1`] = `ShallowWrapper {}`;

exports[`returns isPending correctly 1`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "class": "tc",
    },
    "children": Array [
      Object {
        "data": "Loading",
        "next": null,
        "parent": [Circular],
        "prev": null,
        "type": "text",
      },
    ],
    "name": "h1",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;
